f3/certstore.go

```diff
@@ -0,0 +1,189 @@
+package f3
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/namespace"
+
+	"github.com/Kubuxu/go-broadcast"
+	"golang.org/x/xerrors"
+)
+
+// TODO: setup cbor-gen
+type Cert struct {
+	Instance uint64
+}
+
+func (c *Cert) UnmarshalBinary(b []byte) error {
+	// temp until cbor-gen
+	if len(b) != 8 {
+		return xerrors.Errorf("expected length 8 got %d", len(b))
+	}
+	c.Instance = binary.LittleEndian.Uint64(b)
+	return nil
+}
+func (c Cert) MarshalBinary() ([]byte, error) {
+	return binary.LittleEndian.AppendUint64([]byte{}, c.Instance), nil
+}
+
+var ErrCertAlreadyExists = errors.New("certificate already exists in the CertStore")
+var ErrCertNotFound = errors.New("certificate not found")
+
+// CertStore is responsible for storying and relaying information about new finality certificates
+type CertStore struct {
+	dsLk     sync.Mutex
+	ds       datastore.Datastore
+	busCerts broadcast.Channel[Cert]
+}
+
+func NewCertStore(ctx context.Context, ds datastore.Datastore) (*CertStore, error) {
+	cs := &CertStore{
+		ds: namespace.Wrap(ds, datastore.NewKey("/certstore")),
+	}
+	latestCert, err := cs.loadLatest(ctx)
+	if err != nil {
+		return nil, xerrors.Errorf("loading latest cert: %w", err)
+	}
+	if latestCert != nil {
+		cs.busCerts.Publish(*latestCert)
+	}
+
+	return cs, nil
+}
+
+var certStoreLatestKey = datastore.NewKey("/latestCert")
+
+func (cs *CertStore) loadLatest(ctx context.Context) (*Cert, error) {
+	cb, err := cs.ds.Get(ctx, certStoreLatestKey)
+	if errors.Is(err, datastore.ErrNotFound) {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, xerrors.Errorf("getting latest cert: %w", err)
+	}
+	var c Cert
+	err = c.UnmarshalBinary(cb)
+	if err != nil {
+		return nil, xerrors.Errorf("unmarshalling latest cert: %w", err)
+	}
+	return &c, err
+}
+
+func (cs *CertStore) Latest() *Cert {
+	return cs.busCerts.Last()
+}
+
+func (cs *CertStore) Get(ctx context.Context, instance uint64) (*Cert, error) {
+	cs.dsLk.Lock()
+	b, err := cs.ds.Get(ctx, cs.keyForInstance(instance))
+	cs.dsLk.Unlock()
+
+	if errors.Is(err, datastore.ErrNotFound) {
+		return nil, xerrors.Errorf("cert at %d: %w", instance, ErrCertNotFound)
+	}
+	if err != nil {
+		return nil, xerrors.Errorf("accessing cert in datastore: %w", err)
+	}
+
+	var c Cert
+	if err := c.UnmarshalBinary(b); err != nil {
+		return nil, xerrors.Errorf("unmarshalling cert: %w", err)
+	}
+	return &c, err
+}
+
+// GetRange returns a range of certs from start to end inclusive.
+// If it encouters missing cert, it returns a wrapped ErrCertNotFound and the available certs
+func (cs *CertStore) GetRange(ctx context.Context, start uint64, end uint64) ([]Cert, error) {
```

Comment:
What do start and end represent? Indices? Epoch numbers?
In which order are they traversed? Can they be traversed the other way?

(Please expand docs with all the semantics)